allennlp/modules/span_extractors/bidirectional_endpoint_span_extractor.py

```diff
@@ -0,0 +1,230 @@
+
+    def get_input_dim(self) -> int:
+        return self._input_dim
+
+    def get_output_dim(self) -> int:
+        unidirectional_dim = int(self._input_dim / 2)
+        forward_combined_dim = get_combined_dim(self._forward_combination,
+                                                [unidirectional_dim, unidirectional_dim])
+        backward_combined_dim = get_combined_dim(self._backward_combination,
+                                                 [unidirectional_dim, unidirectional_dim])
+        if self._span_width_embedding is not None:
+            return forward_combined_dim + backward_combined_dim + \
+                   self._span_width_embedding.get_output_dim()
+        return forward_combined_dim + backward_combined_dim
+
+    @overrides
+    def forward(self, # pylint: disable=arguments-differ
+                sequence_tensor: torch.FloatTensor,
+                indices: torch.LongTensor,
+                sequence_mask: torch.LongTensor = None,
+                indices_mask: torch.LongTensor = None) -> torch.FloatTensor:
+
+        # Both of shape (batch_size, sequence_length, embedding_size / 2)
+        forward_sequence, backward_sequence = torch.split(sequence_tensor, int(self._input_dim / 2), -1)
+        forward_sequence = forward_sequence.contiguous()
+        backward_sequence = backward_sequence.contiguous()
+
+        # shape (batch_size, num_spans)
+        span_starts, span_ends = [index.squeeze(-1) for index in indices.split(1, dim=-1)]
+
+        if indices_mask is not None:
+            span_starts *= indices_mask
+            span_ends *= indices_mask
+
+        # We want `exclusive` span starts, so we remove 1 from the forward span starts
+        # as the AllenNLP ``SpanField`` is inclusive.
+        # shape (batch_size, num_spans)
+        exclusive_span_starts = span_starts - 1
+        # shape (batch_size, num_spans)
+        start_sentinel_mask = (exclusive_span_starts == -1).float().unsqueeze(-1)
+
+        # We want `exclusive` span ends for the backward direction
+        # (so that the `start` of the span in that direction is exlusive), so
+        # we add 1 to the span ends as the AllenNLP ``SpanField`` is inclusive.
+        exclusive_span_ends = span_ends + 1
+
+        if sequence_mask is not None:
+            # shape (batch_size)
+            sequence_lengths = get_lengths_from_binary_sequence_mask(sequence_mask)
+        else:
+            # shape (batch_size), filled with the sequence length size of the sequence_tensor.
+            sequence_lengths = ones_like(sequence_tensor[:, 0, 0]).long() * sequence_tensor.size(1)
+
+        # shape (batch_size, num_spans)
+        end_sentinel_mask = (exclusive_span_ends == sequence_lengths.unsqueeze(-1)).long()
+
+        # Forward Direction: start indices are exclusive. Shape (batch_size, num_spans, input_size / 2)
+        forward_start_embeddings = batched_index_select(forward_sequence, exclusive_span_starts)
```

Comment:
How is this not making your tests crash?  Won't you have a `-1` in `exclusive_span_starts` for some instances?  You haven't removed that by this point.  Is `-1` a valid entry for `index_select`?

Reply:
Ah - this is a weird edge case. What's happening here is that the index is being adjusted wrt to the batch, so it actually has some value added to it to make it non-negative. If it was in the first element of the batch it would have crashed. I'll add the mask to be explicit.